src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -146,8 +160,31 @@ private Cci.ITypeReference GetStorageStruct(uint size)
             return new ExplicitSizeStruct(size, this, _systemValueType);
         }
 
+        internal bool HasStaticConstructor => _staticConstructor != null;
+
+        internal Cci.IFieldReference GetOrAddModuleVersionId(Cci.ITypeReference mvidType, Func<Cci.IMethodDefinition> staticConstructorCreator)
+        {
+            if (_mvidField == null)
+            {
+                Interlocked.CompareExchange(ref _mvidField, new ModuleVersionIdField("MVID", this, mvidType), null);
+                EnsureStaticConstructorExists(staticConstructorCreator);
+            }
+
+            return _mvidField;
```

Comment:
Consider asserting that mvidType matches the type of the field. This will ensure all consumers are using the same type.